File: raspberry-pi-ios-client/RaspberryPiClient/Data/Infrastructure/Foxglove/Server/FoxgloveAdvertiseServices.swift
//
//  FoxgloveAdvertiseServices.swift
//  RaspberryPiClient
//
//  Created by Shusuke Ota on 2025/03/20.
//

import Foundation

/// Advertise Services メッセージ（サーバー → クライアント）
/// 利用可能なサービスをクライアントに通知する
/// Ref: https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#advertise-services
struct FoxgloveAdvertiseServices: Codable {
    let op: FoxgloveMessageOperation
    let services: [FoxgloveAdvertiseService]

    init(op: FoxgloveMessageOperation = .advertiseServices, services: [FoxgloveAdvertiseService]) {
        self.op = op
        self.services = services
    }

    /// サービス名 → serviceId のマッピングを返す
    func serviceNameToIdMap() -> [String: UInt32] {
        Dictionary(uniqueKeysWithValues: services.map { ($0.name, $0.id) })
    }

    /// サービス名 → リクエスト encoding のマッピングを返す（advertiseServices の request.encoding）
    func serviceNameToRequestEncodingMap() -> [String: String] {
        Dictionary(uniqueKeysWithValues: services.compactMap { service in
            guard let encoding = service.request?.encoding else { return nil }
            return (service.name, encoding)
        })
    }
}

/// Advertise Services メッセージ内のサービス情報（サーバー → クライアント）
/// Ref: https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#advertise-services
struct FoxgloveAdvertiseService: Codable {
    let id: UInt32
    let name: String
    let type: String?
    let request: FoxgloveServiceSchema?
    let response: FoxgloveServiceSchema?
    let requestSchema: String?
    let responseSchema: String?
}

/// サービスのリクエスト/レスポンススキーマ
struct FoxgloveServiceSchema: Codable {
    let encoding: String?
    let schemaName: String?
    let schemaEncoding: String?
    let schema: String?
}

File: raspberry-pi-android-client/app/src/main/java/com/shusuke/raspberry_pi_client/data/infrastructure/foxglove/server/FoxgloveAdvertiseServices.kt
package com.shusuke.raspberry_pi_client.data.infrastructure.foxglove.server

import com.shusuke.raspberry_pi_client.data.infrastructure.foxglove.common.FoxgloveMessageOperation
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Advertise Services メッセージ（サーバー → クライアント）
 *
 * 利用可能なサービスをクライアントに通知する。
 * Ref: https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#advertise-services
 */
@Serializable
data class FoxgloveAdvertiseServices(
    val op: FoxgloveMessageOperation = FoxgloveMessageOperation.AdvertiseServices,
    val services: List<FoxgloveAdvertiseService> = emptyList(),
) {
    fun serviceNameToIdMap(): Map<String, UInt> =
        services.associate { it.name to it.id }

    fun serviceNameToRequestEncodingMap(): Map<String, String> =
        services.mapNotNull { service ->
            service.request?.encoding?.let { (service.name to it) }
        }.toMap()
}

/**
 * Advertise Services メッセージ内のサービス情報（サーバー → クライアント）
 *
 * Ref: https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#advertise-services
 */
@Serializable
data class FoxgloveAdvertiseService(
    val id: UInt,
    val name: String,
    val type: String? = null,
    val request: FoxgloveServiceSchema? = null,
    val response: FoxgloveServiceSchema? = null,
    @SerialName("requestSchema")
    val requestSchema: String? = null,
    @SerialName("responseSchema")
    val responseSchema: String? = null,
)

/**
 * サービスのリクエスト/レスポンススキーマ
 */
@Serializable
data class FoxgloveServiceSchema(
    val encoding: String? = null,
    @SerialName("schemaName")
    val schemaName: String? = null,
    @SerialName("schemaEncoding")
    val schemaEncoding: String? = null,
    val schema: String? = null,
)
